src/bottom_up_iterators/bottom_up_depth_iterator.jl

```diff
@@ -0,0 +1,200 @@
+Base.@doc """
+    @programiterator BUDepthIterator(problem::Union{Nothing, Problem{Vector{IOExample}}}=nothing, obs_equivalence::Bool=false) <: BottomUpIterator
+
+Implementation of the `BottomUpIterator`. Iterates through complete programs in increasing order of their depth.
+""" BUDepthIterator
+@programiterator BUDepthIterator(
+    problem::Union{Nothing, Problem{Vector{IOExample}}}=nothing,
+    obs_equivalence::Bool=false
+) <: BottomUpIterator
+
+"""
+    struct BUDepthBank <: BottomUpBank
+
+Specialization of `BottomUpState`'s `BottomUpBank` type. Holds a `Dict` mapping `Symbols` to `Vectors` of `RuleNodes` of that `Symbol`.
+"""
+struct BUDepthBank <: BottomUpBank
+    rulenodes_by_symbol::Dict{Symbol, Vector{RuleNode}}
+end
+
+"""
+    struct BUDepthData <: BottomUpData
+
+Specialization of `BottomUpState`'s `BottomUpData` type. Contains the following:
+* `nested_rulenode_iterator::NestedRulenodeIterator`: Iterator generating `RuleNodes` using a grammar rule to combine existing `RuleNodes` from the `bank`.
+* `new_programs::Vector{RuleNode}`: Programs generated at the current depth level. Will be appended to the `bank` when starting a new depth level.
+* `rules::Queue{Int}`: Grammar rules left to be used at the current depth level. 
+* `depth::Int`: Current depth.
+"""
+mutable struct BUDepthData <: BottomUpData
+    nested_rulenode_iterator::NestedRulenodeIterator
+    new_programs::Vector{RuleNode}
+    rules::Queue{Int}
+    depth::Int
+end
+
+"""
+	order(iter::BUDepthIterator)::Queue{Int}
+
+Returns the non-terminal rules in the order in which they appear in the grammar.
+"""
+function order(
```

Comment:
`order` is too vague. I suggest we make the function name more descriptive (`ordered_nonterminals`?). Also, I believe this should be possible with a one-liner `Queue(grammar.rules[grammar.isterminal])` or something like this.